packages/kbn-es/src/cluster.ts

```diff
@@ -559,51 +567,53 @@ exports.Cluster = class Cluster {
     }
   }
 
-  javaOptions(options) {
-    let esJavaOpts = `${options.esJavaOpts || ''} ${process.env.ES_JAVA_OPTS || ''}`;
+  javaOptions(esJavaOpts: string | undefined) {
+    let _esJavaOpts = `${esJavaOpts || ''} ${process.env.ES_JAVA_OPTS || ''}`;
 
     // ES now automatically sets heap size to 50% of the machine's available memory
     // so we need to set it to a smaller size for local dev and CI
     // especially because we currently run many instances of ES on the same machine during CI
     // inital and max must be the same, so we only need to check the max
-    if (!esJavaOpts.includes('Xmx')) {
+    if (!_esJavaOpts.includes('Xmx')) {
       // 1536m === 1.5g
-      esJavaOpts += ' -Xms1536m -Xmx1536m';
+      _esJavaOpts += ' -Xms1536m -Xmx1536m';
     }
-    return esJavaOpts.trim();
+    return _esJavaOpts.trim();
   }
 
   /**
    * Run an Elasticsearch Serverless Docker cluster
-   *
-   * @param {ServerlessOptions} options
+   * @param options ServerlessOptions
    */
-  async runServerless(options = {}) {
-    if (this._process || this._outcome) {
-      throw new Error('ES has already been started');
+  async runServerless(options: ServerlessOptions) {
+    if (this.process) {
+      throw new Error('ES stateful cluster has already been started');
+    }
+
+    if (this.serverlessNodes.length > 0) {
+      throw new Error('ES serverless docker cluster has already been started');
     }
 
-    this._serverlessNodes = await runServerlessCluster(this._log, options);
+    this.serverlessNodes = await runServerlessCluster(this.log, options);
 
     if (options.teardown) {
       /**
        * Ideally would be async and an event like beforeExit or SIGINT,
        * but those events are not being triggered in FTR child process.
        */
-      process.on('exit', () => teardownServerlessClusterSync(this._log, options));
+      process.on('exit', () => teardownServerlessClusterSync(this.log, options));
     }
   }
 
   /**
    * Run an Elasticsearch Docker container
-   *
-   * @param {DockerOptions} options
+   * @param options DockerOptions
    */
-  async runDocker(options = {}) {
-    if (this._process || this._outcome) {
-      throw new Error('ES has already been started');
+  async runDocker(options: DockerOptions) {
+    if (this.process) {
+      throw new Error('ES stateful cluster has already been started');
     }
 
-    this._process = await runDockerContainer(this._log, options);
+    await runDockerContainer(this.log, options);
```

Comment:
`process` property is used only by stateful ES instance started from snapshot and reference the child process (has type of `ExecaChildProcess`)

`await runDockerContainer(...)` resolves the promise and returns the actual result of child process execution, so it has a different meaning and we can't assign it to the propertly.
I don't think we need to store that result at all. Thoughts?

Reply:
I also couldn't really find any good solution to this. The call site really didn't do anything with the child process, because the running docker processes are terminated with another docker command, and not through the PID. 

So I think we can go with not returning the result here. 
(But if we really want to keep the option, you can always wrap the result to an object, so the promise won't wait for the process' end.)

Reply:
> I also couldn't really find any good solution to this. The call site really didn't do anything with the child process, because the running docker processes are terminated with another docker command, and not through the PID.
> 
> So I think we can go with not returning the result here. (But if we really want to keep the option, you can always wrap the result to an object, so the promise won't wait for the process' end.)

Hello there, can you clarify the bit about wrapping the result in an object pretty please?

Reply:
> Hello there, can you clarify the bit about wrapping the result in an object pretty please?
So, the problem comes from the behaviour that `execa`'s return value is this type:
```typescript
type ExecaChildProcess<StdoutErrorType = string> = ChildProcess &
		ExecaChildPromise<StdoutErrorType> &
		Promise<ExecaReturnValue<StdoutErrorType>>;
```
; note the `&` and not a `|` as you'd normally expect. So the return type is a `Promise<...<string>>` AND a `ChildProcess`. The promise is to resolve when the process ends. However, we're aiming to do something with the `ChildProcess` (e.g.: inspect PID, read stdout stream). 

The problem comes from the fact, that if you're retuning any `Promise` in a `.then` clause or an async function, it's flatmapped. You won't be able to wrap `ExecaChildProcess` into a promise, then, because it's already of a promise type.

So you won't be able to construct a `Promise<ExecaChildProcess>`, but only `ExecaChildProcess`. As soon as you `await` on, or use `.then` on this, you'll inevitably wait for the child process to terminate, and have its stdout available, which is not what we wanted to be similar to the other functions.

Solution:
You can return `Promise<{ x: ExecaChildProcess }>` from your async function, this way, when you `await` on the result, you'll just wait for any of the awaits in the function, but not to the end of the child process. You can then separately `await` the `ExecaChildProcess` if you want, or use it as a ChildProcess even before the process completes. 

```
// return like this
return {
    x: execa('docker', dockerCmd, {
      // inherit is required to show Docker output and Java console output for pw, enrollment token, etc
      stdio: ['ignore', 'inherit', 'inherit'],
    })
  }

// use like this
this._process = (await runDockerContainer(this._log, options)).x;
```

Reply:
Thanks for the detailed explanation, Alex. I agree that we can technically assign it to the `process` property, but I believe we should keep using `process` only for stateful not-from-docker start.

Afaik `runDockerContainer()` is resolved only on process exit and I don't see any on-exit logic that might be using `process` to stop container or something similar. Or I miss something?

Reply:
I also didn't see any use of the returned values. The PIDs here are not needed for terminating the docker images, as they don't run as processes. I think it's good as you did it, not to expose misleading properties.